src/test/scala/io/iohk/ethereum/mpt/MerklePatriciaTrieSuite.scala

```diff
@@ -0,0 +1,619 @@
+    val key4: Array[Byte] = Hex.decode("123500")
+    val trie = EmptyTrie.put(key1, key1).put(key2, key2).put(key3, key3).put(key4, key4)
+    val wrongSource = HashMapDataSource().update(
+      rootHash = Array.emptyByteArray,
+      toRemove = Seq(),
+      toUpdate = Seq(trie.getRootHash -> trie.dataSource.get(trie.getRootHash).get)
+    )
+    val trieWithWrongSource = MerklePatriciaTrie[Array[Byte], Array[Byte]](trie.getRootHash, wrongSource, hashFn)
+    val trieAfterDelete = Try {
+      trieWithWrongSource.remove(key1)
+    }
+    assert(trieAfterDelete.isFailure)
+  }
+
+  test("Get in an empty trie"){
+    val obtained = EmptyTrie.get(Hex.decode("1234"))
+    assert(obtained.isEmpty)
+  }
+
+  test("Insert with an empty key in a branch node"){
+    val key1: Array[Byte] = Hex.decode("00")
+    val key2: Array[Byte] = Hex.decode("f0")
+    val trieWithBranch = EmptyTrie.put(key1, key1).put(key2, key2)
+    val key3: Array[Byte] = Hex.decode("")
+    val val3: Array[Byte] = Hex.decode("1234")
+    val trie = trieWithBranch.put(key3, val3)
+    val key1Get = trie.get(key1)
+    val key2Get = trie.get(key2)
+    val key3Get = trie.get(key3)
+    assert(key1Get.isDefined && (key1Get.get sameElements key1))
+    assert(key2Get.isDefined && (key2Get.get sameElements key2))
+    assert(key3Get.isDefined && (key3Get.get sameElements val3))
+  }
+
+  test("Remove of a key (not in trie) whose value should be in a branch node"){
+    val key1: Array[Byte] = Hex.decode("1100")
+    val key2: Array[Byte] = Hex.decode("11f0")
+    val trieWithBranch = EmptyTrie.put(key1, key1).put(key2, key2)
+    val trieAfterDelete = trieWithBranch.remove(Hex.decode("11"))
+    assert(trieAfterDelete.getRootHash sameElements trieWithBranch.getRootHash)
+  }
+
+  test("Remove of a key (not in trie) that should be in the child of a BranchNode that is not present"){
+    val key1: Array[Byte] = Hex.decode("1100")
+    val key2: Array[Byte] = Hex.decode("11f0")
+    val trieWithBranch = EmptyTrie.put(key1, key1).put(key2, key2)
+    val trieAfterDelete = trieWithBranch.remove(Hex.decode("11a0"))
+    assert(trieAfterDelete.getRootHash sameElements trieWithBranch.getRootHash)
+  }
+
+  /* IODB tests */
+  test("Simple test with IODB as Source") {
+    //create temporary dir
+    val dir = File.createTempFile("iodb", "iodb")
+    dir.delete()
+    dir.mkdir()
+
+    //open new store
+    val dataSource = new IodbDataSource(new LSMStore(dir = dir, keySize = 32))
+    val emptyTrie = MerklePatriciaTrie[Int, Int](dataSource, hashFn)
+    val trieWithOneElement = emptyTrie.put(1, 5)
+    val obtained = trieWithOneElement.get(1)
+    assert(obtained.isDefined)
+    assert(obtained.get == 5)
+    val trieAfterDelete = trieWithOneElement.remove(1)
+    val obtainedAfterDelete = trieAfterDelete.get(1)
+    assert(obtainedAfterDelete.isEmpty)
+  }
+
+  ignore("IODB test - Insert of the first 5000 numbers hashed and then remove half of them"){
```

Comment:
I have several remarks about those tests:

1. If we're ignoring tests in such manner it deserves at least a comment. Otherwise an unknowing developer might come at those tests and wonder what he has to do to fix them.
2. For such long running tests I think a better solution would be to create them in a separate test configuration. Then in Circle CI we could easily run them conditionally with `sbt long-running:test`, e.g. only when merging to master.
  Another reason for a separate config is that, unless I'm missing something, they don't look like regular unit tests. Are they stress tests? Does running them with 40000 numbers give us a big boost of confidence as compared to running with 4 numbers?
  Lastly, if we don't run those tests automatically how are we going maintain them? Are we going to nominate a person who's going to run them occasionally to see if they're not broken? I certainly do not volunteer 😁 
3. I ran one of those tests: `"IODB Test - PatriciaTrie insert and get"` and it takes ~600 ms. I wouldn't say that's terribly slow.
4. Could we test `IodbDataSource` in a separate suite and without dependency on `MerklePatriciaTrie`? It is my understanding MPT depends on a DataSource, not vice versa. More importantly though, what if we find out that IODB doesn't meet our expectations? We shouldn't have to change MPT tests when we replace the DataSource implementation.